### DISCUSSION TITLE
Add other types of filters, like select, checkbox, and radio as display option for Nested Tabs, Taxonomy Filters, and Gallery Filters

### DISCUSSION THREAD

**Discussion by Henzxon** [elementor/elementor#23364] [Feature Request]
### Prerequisites

- [X] I have searched for similar features requests in both open and closed tickets and cannot find a duplicate.
- [X] The feature is still missing in the latest stable version of Elementor ( Elementor Pro )

### What problem is your feature request going to solve? Please describe.

> _Tabs consumes way too much space and scrolling through tabs might not be an  immediate action on most users._

### Describe the solution you'd like

> _It would be great if a dropdown selection would be available for tabs, or any category or taxonomy filter. This would make it easier to navigate a long list of taxonomy or category filter when horizontal scrolling is not ideal._

### Describe alternatives you've considered

> _Sad to say but had to resort to plugin..._

**Comment by nicholaszein** [comment]
Hello @Henzxon,

Thanks for contributing your suggestion! 🙏

Your ideas and input as a Web Creator are valuable to us. We're always keen to explore ways to enhance Elementor to meet our community's needs better.

### 💡 Due to the high volume of feature requests we receive and our commitment to continuous improvement, we can't guarantee that every suggestion will be immediately included in our roadmap.

That being said, your idea is important to us. We'll keep your suggestion on our radar, and if it attracts a significant number of votes in the future, we'll consider implementing it.

We appreciate your patience as we take the time to review your request thoroughly. We aim to ensure all decisions align with our mission to provide Web Creators with the tools they need to build their dream websites.

Kind regards

**Reply by sayeedbaksh** [reply]
ecommerce features are so lacking, i subscribe to astra pro just do make my site look good. i spend too much time messing with astra, using css. id rather buy elementor widgets.

**Comment by esuarkSunil** [comment]
Hello :)

for a new Website I really need this feature. I don't want to use any more plugins for this.
Are there any new informations for this feature request?

Tank you for your work!

**Comment by davelavoie** [comment]
For those who still need to use a `<select>` instead of the default filters, here's a JavaScript alternative until Elementor supports it by default. It has been developed for the Taxonomy Filter, so it might need some adjustments for other types of filters.

1. First, make sure to add an ID to the Taxonomy Filter widget (Advanded -> CSS ID).

2. Then, add an HTML widget just after the filter(s) you created, and insert the following code:
```
<script>
jQuery(function() {
	function generateSelect( id ) {
	// Gets all filters.
		var $options = jQuery( '#' + id + ' .e-filter');

		// Hides the default Elementor filters.
		$options.hide();

		// Reorders filters so the "All" option is the first one.
		$options.find('button[data-filter="__all"]').prependTo( $options );

		// Creates the select.
		jQuery('<select id="select-' + id + '" />').appendTo( '#' + id + ' .elementor-widget-container' );

		// Loops through the filters and adds them to the select.
		$options.find('button').each(function() {
			var filter = jQuery(this);
			jQuery("<option />", {
				"value": filter.attr("data-filter"),
				"text": filter.text()
			}).appendTo('#select-' + id );
		});

		// Adds an event to the select so it triggers a click to the appropriate filter when an option is selected.
		jQuery('#select-' + id).change(function() {
			var selectedFilter = jQuery(this).find("option:selected").val();
			jQuery('#' + id + ' .e-filter button[data-filter="' + selectedFilter + '"').click();
		});
	}

	// Generates the select for the following Taxonomy filters.
	generateSelect( 'filter-id-1' );
	generateSelect( 'filter-id-2' );
});
</script>
```

3. Near the end of the script above, after the `// Generates the select for the following Taxonomy filters.` line, simply edit the parameter of the `generateSelect()` functions and replace the values with the IDs you added to the taxonomy filter during step 1. Obviously if you only need one filter, remove the second `generateSelect()` function.

Hope it helps! Feel free to share any improvement to this code so others can benefit from this solution!

**Reply by arnasofc** [reply]
Thank you very much! It works perfectly.

**Reply by johnnny-bravoo** [reply]
Thank you so much @davelavoie for your great help. You're a super genuis. 

For those using [jQuery Nice Select](https://jqueryniceselect.hernansartorio.com/) who notice the dropdown staying open after selecting a taxonomy, I’ve got a simple fix. I spent quite some time figuring out why it wouldn’t close properly, so hopefully this saves you the trouble.

```
// Inside your generateSelect()
jQuery(document).on('click', '#'+id+' .nice-select.open ul.list li', function(e) {
    var $niceSelect = jQuery(this).closest('.nice-select');
    setTimeout(function(){
        $niceSelect.removeClass('open'); 
    }, 100); // small delay so it doesn’t instantly reopen
});

```

**Comment by nilssondigital** [comment]
Second this...this would be awesome and expands the usability to create a directory-style page, and cuts out the need for additional plugins which either charge for the feature, or are included in a full add-on suite with 100 other widgets that aren't needed.

The snippet above works amazing for select/dropdown option as well so much thanks @davelavoie  - a radio/checkbox snippet would be incredible if anyone is smart + generous enough + has the time!

**Comment by nirjhareswar** [comment]
For those who wants a solution for checkbox, multiple.

I broke in two parts. One "clear all" button / text to uncheck all checked boxes.

sorm-filter-domains is the css id

```
<div class="cbd-sorm-filter-domains-cls"><span class="cb-sorm-filter-domains-cls" id="cb-sorm-filter-domains-0" name="cb-sorm-filter-domains-0" value="__all" data-gaconnector-tracked="true">Clear All</span></div>

<script>
jQuery(function() {	
	// add for clear all
 	jQuery('#cb-sorm-filter-domains-0').click(function() {
		jQuery('#sorm-filter-domains .e-filter button[data-filter="__all"').click();
		jQuery( ".cb-sorm-filter-domains-cls" ).prop( "checked", false );
	});
});
</script>
```

Second part is the function itself.

```
	function generateCheckbox( id ) {
	// Gets all filters.
		var $options = jQuery( '#' + id + ' .e-filter');

		// Hides the default Elementor filters.
		$options.hide();

		// Reorders filters so the "All" option is the first one.
		$options.find('button[data-filter="__all"]').prependTo( $options );

		// Loops through the filters and adds them as checkbox
		$options.find('button').each(function(index, element) {
			if(index != 0) {	// All we will use "Clear All"
				var filter = jQuery(this);
			
				ne = jQuery('<div />', { class: 'cbd-'+id+'-cls' });
				ne.append(jQuery('<input />', { type: 'checkbox', class: 'cb-'+id+'-cls', id: 'cb-'+id+'-'+index, name: 'cb-'+id+'-'+index, value: filter.attr("data-filter") }));
				ne.append(jQuery('<label />', { 'for': 'cb-'+id+'-'+index, class: 'cbl-'+id+'-cls', text: filter.text() }));
				ne.appendTo( '#' + id + ' .elementor-widget-container' );
			}	
		});

		// Adds an event to the checkbox so it triggers a click to the appropriate filter when checked or unchecked
		jQuery('.cb-'+id+'-cls').change(function() {
			var selectedFilter = jQuery(this).val();		
			jQuery('#' + id + ' .e-filter button[data-filter="' + selectedFilter + '"').click();
		});
		
	}

// Generates the checkbox for the following Taxonomy filters.
	generateCheckbox( 'sorm-filter-domains' );

```


Self explanatory, and extended @davelavoie s function. @nilssondigital you can extend further for radio as well similarly.

Cheers!

**Reply by nilssondigital** [reply]
@nirjhareswar this is amazing thank you! Worked like a charm.

There's good people in the world and you are one of them! Thank you so much!

**Reply by nirjhareswar** [reply]
@nilssondigital You are so kind! 🙏

**Comment by Occupy-Site-Editor** [comment]
Hey @davelavoie - thanks heaps! This made such a difference to the site I'm working on. My skills are fairly low level, so I'm now struggling to see how to add any styling to the dropdown filter options. A white background is the main thing I need, instead of the default grey. Update: Got it.. filter-id-1-my-id {background-color: white; }

**Comment by RubenMartins** [comment]
Another approach to when you are dealing with multiple taxonomy filters you can just set a id to the parent container and set it here.

PS. This works as well for the "Optimized Markup" feature enabled from Elementor

```
jQuery(function() {
	var container_id = 'CONTAINER_ID_HERE';

	// Check if the container exists.
	if( jQuery('#' + container_id).length > 0){
		function generateSelect( elem ) {
			// Gets all filters.
			var $options = elem.find('.e-filter');
			var id = elem.data('id');

			// Hides the default Elementor filters.
			$options.hide();

			// Reorders filters so the "All" option is the first one.
			$options.find('button[data-filter="__all"]').prependTo( $options );

			// (OPTIONAL) Creates the label.
			var title = $options.find('button[data-filter="__all"]').text();
			jQuery('<label for="select-' + id + '">' + title + '</label>').appendTo( elem );

			// Creates the select.
			jQuery('<select id="select-' + id + '" />').appendTo( elem );

			// Loops through the filters and adds them to the select.
			$options.find('button').each(function() {
				var filter = jQuery(this);
				jQuery("<option />", {
					"value": filter.attr("data-filter"),
					"text": filter.text(),
					selected: filter.attr("aria-pressed") === "true"
				}).appendTo('#select-' + id );
			});

			// Adds an event to the select so it triggers a click to the appropriate filter when an option is selected.
			jQuery('#select-' + id).change(function() {
				var selectedFilter = jQuery(this).find("option:selected").val();
				elem.find('.e-filter button[data-filter="' + selectedFilter + '"]').click();
			});
		}

		// loop through all the filters and add a select to each one.
		jQuery('#' + container_id + ' .elementor-widget-taxonomy-filter').each(function() {
			var elem = jQuery(this);
			if (elem) {
				generateSelect(elem);
			}
		});
	}
});
```

**Comment by JennyPower** [comment]
Taxonomy Filter is great but it's very limiting without having dropdowns

**Reply by chrisp2u** [reply]
Yes, having some additional display options (check box, dropdown) for the taxonomy filter(s) would be nice. I really can't believe that this hasn't been added since this feature request (almost 2 years ago). The code shared above works great, but having these options built-in would be far more convenient.
